File: C#/Combinatorics/Combinatorics.cs
/*
 * Combinatorics.cs
 * Author: Андрій Будильников
 * 
 * Basic combinatorics functions implementation
 */

using System;

namespace Combinatorics
{
    public class CombinatoricsCalculator
    {
        /// <summary>
        /// calculate factorial of n
        /// </summary>
        /// <param name="n">non-negative integer</param>
        /// <returns>factorial of n</returns>
        public static long Factorial(int n)
        {
            if (n < 0)
                throw new ArgumentException("factorial is not defined for negative numbers");
                
            if (n == 0 || n == 1)
                return 1;
                
            long result = 1;
            for (int i = 2; i <= n; i++)
            {
                result *= i;
            }
            return result;
        }

        /// <summary>
        /// calculate permutations p(n,r) = n!/(n-r)!
        /// </summary>
        /// <param name="n">total number of items</param>
        /// <param name="r">number of items to choose</param>
        /// <returns>number of permutations</returns>
        public static long Permutation(int n, int r)
        {
            if (n < 0 || r < 0)
                throw new ArgumentException("n and r must be non-negative");
                
            if (r > n)
                throw new ArgumentException("r cannot be greater than n");
                
            return Factorial(n) / Factorial(n - r);
        }

        /// <summary>
        /// calculate combinations c(n,r) = n!/((n-r)!*r!)
        /// </summary>
        /// <param name="n">total number of items</param>
        /// <param name="r">number of items to choose</param>
        /// <returns>number of combinations</returns>
        public static long Combination(int n, int r)
        {
            if (n < 0 || r < 0)
                throw new ArgumentException("n and r must be non-negative");
                
            if (r > n)
                throw new ArgumentException("r cannot be greater than n");
                
            return Factorial(n) / (Factorial(n - r) * Factorial(r));
        }

        /// <summary>
        /// calculate nth catalan number
        /// </summary>
        /// <param name="n">index of catalan number</param>
        /// <returns>nth catalan number</returns>
        public static long CatalanNumber(int n)
        {
            if (n < 0)
                throw new ArgumentException("n must be non-negative");
                
            return Combination(2 * n, n) / (n + 1);
        }
    }
}

File: Python/Combinatorics/combinatorics.py
"""
combinatorics.py
Author: Андрій Будильников

Basic combinatorics functions implementation
"""

import math

def factorial(n):
    """calculate factorial of n"""
    if n < 0:
        raise ValueError("factorial is not defined for negative numbers")
    if n == 0 or n == 1:
        return 1
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result

def permutation(n, r):
    """calculate permutations p(n,r) = n!/(n-r)!"""
    if n < 0 or r < 0:
        raise ValueError("n and r must be non-negative")
    if r > n:
        raise ValueError("r cannot be greater than n")
    return factorial(n) // factorial(n - r)

def combination(n, r):
    """calculate combinations c(n,r) = n!/((n-r)!*r!)"""
    if n < 0 or r < 0:
        raise ValueError("n and r must be non-negative")
    if r > n:
        raise ValueError("r cannot be greater than n")
    return factorial(n) // (factorial(n - r) * factorial(r))

def catalan_number(n):
    """calculate nth catalan number"""
    if n < 0:
        raise ValueError("n must be non-negative")
    return combination(2 * n, n) // (n + 1)
